indico/modules/rb_new/operations/bookings.py

```diff
@@ -209,3 +214,47 @@ def get_booking_occurrences(booking):
     date_range = sorted(set(cand.start_dt.date() for cand in booking.occurrences))
     occurrences = group_by_occurrence_date(booking.occurrences)
     return date_range, occurrences
+
+
+def check_room_available(room, start_dt, end_dt):
+    occurrences = get_existing_room_occurrences(room, start_dt, end_dt, allow_overlapping=True)
+    prebookings = [occ for occ in occurrences if not occ.reservation.is_accepted]
+    bookings = [occ for occ in occurrences if occ.reservation.is_accepted]
+    unbookable_hours = get_rooms_unbookable_hours([room])
+    hours_overlap = [hours for hours in unbookable_hours
+                     if overlaps((start_dt.time(), end_dt.time()), (hours.start_time, hours.end_time))]
+    nonbookable_periods = get_rooms_nonbookable_periods([room], start_dt, end_dt)
+    blockings = get_rooms_blockings([room], start_dt, end_dt).get(room.id, [])
+    blocked_for_user = [blocking for blocking in blockings
+                        if not blocking.blocking.can_be_overridden(session.user, room)]
+    user_bookings = [booking for booking in bookings if booking.reservation.booked_for_id == session.user.id]
+    user_prebookings = [prebooking for prebooking in prebookings
+                        if prebooking.reservation.booked_for_id == session.user.id]
+
+    return {
+        'can_book': room.can_be_booked(session.user),
+        'can_prebook': room.can_be_prebooked(session.user),
+        'conflict_booking': bool(bookings),
+        'conflict_prebooking': bool(prebookings),
+        'unbookable': bool(hours_overlap or nonbookable_periods or blocked_for_user),
+        'user_booking': bool(user_bookings),
+        'user_prebooking': bool(user_prebookings),
+    }
+
+
+def create_booking_for_event(room_id, event):
+    try:
+        room = Room.get_one(room_id)
+        default_timezone = timezone(config.DEFAULT_TIMEZONE)
+        start_dt = event.start_dt.astimezone(default_timezone).replace(tzinfo=None)
+        end_dt = event.end_dt.astimezone(default_timezone).replace(tzinfo=None)
+        booking_reason = "Event '%s'" % (event.title)
+        data = dict(start_dt=start_dt, end_dt=end_dt, booked_for_user=event.creator, booking_reason=booking_reason,
+                    repeat_frequency=RepeatFrequency.NEVER, event_id=event.id)
+        resv = Reservation.create_from_data(room, data, session.user)
+        return resv
+    except NoReportError as e:
+        if 'Reservation has no valid occurrences' == e:
```

Comment:
![image](https://user-images.githubusercontent.com/179599/49648715-31bdf000-fa27-11e8-917e-f048ecda40db.png)

We don't do Yoda conditions :)

Also, that string is translated when the error is raised, so it won't work properly in other locales. Do we get any other NoReport errors there? If not let's just not check the reason at all.

Reply:
hahaha, Yoda conditions : D
I agree it's not perfect this way. There's another NoReport error if user is not authorized to book the room

Reply:
But we check that before and don't even allow selection to book, right?

Reply:
Yes, we do in the frontend, we also don't allow selection to book if room is occupied. But I figured that there's some chance somebody will book the room in the meantime and it would be good to show some error. But yeah, I guess it won't happen that somebody looses the rights to book the room in the meantime

Reply:
Yaeh, enough of an edge case that a slightly incorrect error message is fine. People are used to almost-useless error messages from mobile apps etc anyway :/